File: src/js/matrixCalculator.js
"use strict";
const results = require("./results");

class CurrentHolder {
    constructor(row, column, side) {
        this.row = this.startRow = row;
        this.column = this.startColumn = column;
        this.side = side;
    }

    actual(value) {
        if (value < 0)
            return value + this.side;
        if (value >= this.side)
            return value - this.side;
        return value;
    }

    get actualRow() { return this.actual(this.row); }
    get actualColumn() { return this.actual(this.column); }

    next(count) {
        if (count % this.side != 0) {
            this.row++;
            this.column++;
        }
        else {
            this.row = ++this.startRow;
            this.column = --this.startColumn;
        }
    }
}

let createEmptyMatrix = (side) => {
    var result = new Array(side);

    for (var rowIndex = 0; rowIndex < side; rowIndex++)
        result[rowIndex] = new Array(side);

    return result;
};

let testResult = (matrix, expectedSum, side) => {
    for (var rowIndex = 0; rowIndex < side; rowIndex++) {
        var sum = 0;
        for (var columnIndex = 0; columnIndex < side; columnIndex++)
            sum += matrix[rowIndex][columnIndex];

        if (sum != expectedSum)
            return false;
    }

    for (var columnIndex = 0; columnIndex < side; columnIndex++) {
        var sum = 0;
        for (var rowIndex = 0; rowIndex < side; rowIndex++)
            sum += matrix[rowIndex][columnIndex];

        if (sum != expectedSum)
            return false;
    }

    var diagonalSum = 0;
    for (var diagonalIndex = 0; diagonalIndex < side; diagonalIndex++)
        diagonalSum += matrix[diagonalIndex][diagonalIndex];

    if (diagonalSum != expectedSum)
        return false;

    diagonalSum = 0;
    for (var diagonalIndex = 0; diagonalIndex < side; diagonalIndex++)
        diagonalSum += matrix[diagonalIndex][side - diagonalIndex - 1];

    return diagonalSum == expectedSum;
}

module.exports = (side) => {
    if (side % 2 !== 1)
        throw new Error("Side needs to be an odd number.");

    var expectedSum = side * (side * side + 1) / 2;
    var matrix = createEmptyMatrix(side);
    var current = new CurrentHolder(-(side - 1) / 2, Math.floor(side / 2), side);

    for (var count = 1; count <= side * side; count++) {
        matrix[current.actualRow][current.actualColumn] = count;

        current.next(count);
    }

    return new results.MatrixResult(matrix, expectedSum, side, testResult(matrix, expectedSum, side));
}

File: src/rs/src/matrix_calculator.rs
use crate::results;

struct CurrentHolder {
    row: i32,
    column: i32,
    start_row: i32,
    start_column: i32,
    side: i32,
}

impl CurrentHolder {
    fn new(row: i32, column: i32, side: i32) -> CurrentHolder {
        return CurrentHolder {
            row,
            column,
            start_row: row,
            start_column: column,
            side,
        };
    }

    fn actual(&self, value: i32) -> usize {
        return match value {
            v if v < 0 => (v + self.side) as usize,
            v if v >= self.side => (v - self.side) as usize,
            _ => value as usize,
        };
    }

    fn actual_row(&self) -> usize {
        return self.actual(self.row);
    }

    fn actual_column(&self) -> usize {
        return self.actual(self.column);
    }

    fn next(&mut self, count: i32) {
        if count % self.side != 0 {
            self.row += 1;
            self.column += 1;
        } else {
            self.start_row += 1;
            self.row = self.start_row;
            self.start_column -= 1;
            self.column = self.start_column;
        }
    }
}

fn create_empty_matrix(side: i32) -> Vec<Vec<i32>> {
    let mut result = vec![];
    for _ in 0..side {
        result.push(vec![0; side as usize]);
    }
    return result;
}

fn test_result(matrix: &Vec<Vec<i32>>, expected_sum: i32, side: usize) -> bool {
    let mut sum;
    for row_index in 0..side {
        sum = 0;
        for column_index in 0..side {
            sum += matrix[row_index][column_index];
        }
        if sum != expected_sum {
            return false;
        }
    }

    for column_index in 0..side {
        sum = 0;
        for row_index in 0..side {
            sum += matrix[row_index][column_index];
        }
        if sum != expected_sum {
            return false;
        }
    }

    sum = 0;
    for diagonal_index in 0..side {
        sum += matrix[diagonal_index][diagonal_index];
    }
    return sum == expected_sum;
}

pub fn calculate(side: i32) -> results::MatrixResult {
    if side % 2 != 1 {
        return Err("Side needs to be an odd number.".to_owned());
    }

    let expected_sum = side * (side * side + 1) / 2;
    let mut matrix = create_empty_matrix(side);
    let mut current = CurrentHolder::new(-(side - 1) / 2, side / 2, side);

    for count in 1..=(side * side) {
        matrix[current.actual_row()][current.actual_column()] = count;
        current.next(count);
    }
    if test_result(&matrix, expected_sum, side as usize) {
        return Ok(results::MatrixData {
            matrix,
            sum: expected_sum,
            side: side as usize,
        });
    }

    return Err("Calculation error.".to_owned());
}
